compiler/sem/semstmts.nim

```diff
@@ -1648,28 +1650,29 @@ proc semFor(c: PContext, n: PNode; flags: TExprFlags): PNode =
   if isCallExpr and call[0].kind == nkSym and
       call[0].sym.magic in {mFields, mFieldPairs}:
     result = semForFields(c, n, call[0].sym.magic)
-  elif isCallExpr and isClosureIterator(call[0].typ.skipTypes(abstractInst)):
-    # first class iterator:
-    result = semForVars(c, n, flags)
-  elif not isCallExpr or call[0].kind != nkSym or
-      call[0].sym.kind != skIterator:
-    if n.len == 3:
-      n[^2] = implicitIterator(c, "items", n[^2])
-    elif n.len == 4:
-      n[^2] = implicitIterator(c, "pairs", n[^2])
-    else:
-      localReport(c.config, n[^2], reportSem(rsemForExpectsIterator))
-    result = semForVars(c, n, flags)
   else:
+    if isCallExpr and isClosureIterator(call[0].typ.skipTypes(abstractInst)):
+      # first class iterator:
+      discard
+    elif not isCallExpr or call[0].kind != nkSym or
+        call[0].sym.kind != skIterator:
+      if n.len == 3:
+        n[^2] = implicitIterator(c, "items", n[^2])
+      elif n.len == 4:
+        n[^2] = implicitIterator(c, "pairs", n[^2])
+      else:
+        n[^2] = c.config.newError(n[^2], PAstDiag(kind: adSemForExpectedIterator))
+      hasError = n[^2].isError or hasError
     result = semForVars(c, n, flags)
-  if result.kind == nkError:
-    discard # do nothing
+  if hasError or result.kind == nkError:
+    hasError = true
   elif n[^1].typ == c.enforceVoidContext:
     # propagate any enforced VoidContext:
     result.typ = c.enforceVoidContext
   elif efInTypeof in flags:
     result.typ = result.lastSon.typ
   closeScope(c)
+  if hasError: result = c.config.wrapError(result)
```

Comment:
```suggestion
  if result.kind != nkError and hasError:
    result = c.config.wrapError(result)
```

otherwise we end up with unnecessary wrapping of errors.